### PR TITLE
Implement non-blocking select statements

### PR DESCRIPTION
**Pull request by aykevl**
Non-blocking selects are much easier to implement, so let's do that first. It should get a bit more code to compile as well (but not necessarily to run well). It depends on #401 because there would be merge conflicts otherwise.

And yes, this is also part of #285 because the time package contains a select that used to be not compiled but will be with that refactor.

### CODE REVIEW

**Comment by deadprogram** [comment]
Ready for conflict resolution, please.

**Comment by aykevl** [comment]
Rebased.

File: src/runtime/chan.go [hunk 1]
```diff
@@ -143,3 +154,65 @@ func chanClose(ch *channel, size uintptr) {
 		ch.state = chanStateClosed
 	}
 }
+
+// chanSelect is the runtime implementation of the select statement. This is
+// perhaps the most complicated statement in the Go spec. It returns the
+// selected index and the 'comma-ok' value.
+//
+// TODO: do this in a round-robin fashin (as specified in the Go spec) instead
```
Comment (deadprogram) [inline src/runtime/chan.go]: fashion

File: src/runtime/chan.go [hunk 2]
```diff
@@ -143,3 +154,65 @@ func chanClose(ch *channel, size uintptr) {
...
+	for i, state := range states {
+		if state.ch == nil {
+			// A nil channel blocks forever, so don't consider it here.
+			continue
+		}
+		if state.value == nil {
+			// A receive operation.
+			switch state.ch.state {
+			case chanStateSend:
+				println("  state send")
```
Comment (deadprogram) [inline src/runtime/chan.go]: You probably did not mean to leave this here...

File: src/runtime/chan.go [hunk 3]
```diff
@@ -143,3 +154,65 @@ func chanClose(ch *channel, size uintptr) {
...
+				memcpy(recvbuf, senderPromise.ptr, uintptr(state.ch.elementSize))
+				state.ch.blocked = senderPromise.next
+				senderPromise.next = nil
+				activateTask(sender)
+				if state.ch.blocked == nil {
+					state.ch.state = chanStateEmpty
+				}
+				return uintptr(i), true // commaOk = true
+			case chanStateClosed:
+				println("  state closed")
```
Comment (deadprogram) [inline src/runtime/chan.go]: Same as above.

File: src/runtime/chan.go [hunk 4]
```diff
@@ -28,24 +28,35 @@ import (
...
+	elementSize uint16 // the size of one value in this channel
+	state       chanState
+	blocked     *coroutine
 }
 
+type chanState uint8
+
 const (
-	chanStateEmpty = iota
+	chanStateEmpty chanState = iota
```
Comment (deadprogram) [inline src/runtime/chan.go]: :+1:

**Comment by aykevl** [comment]
Oh that was really sloppy of me, thanks for checking. Fixed now.

**Comment by deadprogram** [comment]
IMO this is a very big leap forward in one of the few missing language features. Great work!

**Pull request merged**